### DISCUSSION TITLE
IaC Core Team Meeting 2024-11-19

### DISCUSSION THREAD

**Discussion by Zaid-Ajaj** [pulumi/pulumi#17813] [Meeting Notes]
Hey Pulumi community!

These are the notes from this week's IaC core team meeting as it relates to https://github.com/pulumi/pulumi. We're publishing these as part of our efforts to provide transparency about the decisions we make and how we reach them.

----

### Introducing _deferred_ outputs for language SDKs

Addressing https://github.com/pulumi/pulumi/issues/13581 requires a new construct in language SDKs that allows creating so-called deferred outputs: an output that can be initialized but its value resolved/assigned later down the line in the program. It is needed to break circular references between component resources where although the component resources can depend directly on each other, their underlying resources cannot. 

Although usage of deferred outputs is not common at all in Pulumi programs, they will help convert terraform code which uses circular dependencies between terraform modules (component resources in Pulumi). 

Their API is proposed as follows:

### Python
```python
bucket_arn, resolve = pulumi.deferred_output()
bucket = aws.s3.Bucket("example")
resolve(bucket.arn)
```

### TypeScript
```ts
const [bucketArn, resolveBucketArn] = pulumi.deferredOutput<string>();
const bucket = new aws.s3.Bucket("example", { });
resolveBucketArn(bucket.arn);
```

### .NET (C#)
```csharp
var deferredValue = new DeferredOutput<int>()
var componentA = MakeComponentA(deferredValue.Output);
var componentB = MakeComponentB(componentA.Result);
deferredValue.Set(componentB.Result);
```

### Implementation
- Python + nodejs https://github.com/pulumi/pulumi/pull/17793
- .NET https://github.com/pulumi/pulumi-dotnet/pull/385 
- Go TODO https://github.com/pulumi/pulumi/issues/17788
- Java TODO

**Comment by saiabishek1** [comment]
This is great! When will the docs be updated?
